File: fixtures/ts-basic/src/services/userService.ts
import { UserRepository } from "../repositories/userRepository";
import { EmailService, welcomeEmail } from "./emailService";
import { CreateUserInput, User, isAdmin } from "../models/user";
import { assertValidEmail, isNonEmpty, ValidationError } from "../utils/validation";

function hashPassword(password: string): string {
  // placeholder — a real implementation would use bcrypt/argon2
  return `hashed:${password}`;
}

export class UserService {
  constructor(
    private repo: UserRepository,
    private emailService: EmailService
  ) {}

  register(input: CreateUserInput): User {
    assertValidEmail(input.email);
    if (!isNonEmpty(input.name)) {
      throw new ValidationError("name", "name is required");
    }
    if (this.repo.findByEmail(input.email)) {
      throw new ValidationError("email", "email already registered");
    }

    const user = this.repo.insert(input, hashPassword(input.password));
    this.emailService.send(welcomeEmail(user.name, user.email));
    return user;
  }

  getById(id: string): User | undefined {
    return this.repo.findById(id);
  }

  promoteToAdmin(id: string): User {
    const user = this.repo.findById(id);
    if (!user) throw new ValidationError("id", "user not found");
    if (isAdmin(user)) return user;
    user.role = "admin";
    return user;
  }

  listAll(): User[] {
    return this.repo.all();
  }
}

File: fixtures/csharp-basic/Services/UserService.cs
using OrderSystem.Models;
using OrderSystem.Repositories;
using OrderSystem.Utils;

namespace OrderSystem.Services;

public class UserService
{
    private readonly IUserRepository _userRepository;

    public UserService(IUserRepository userRepository)
    {
        _userRepository = userRepository;
    }

    public User Register(int id, string email, string name)
    {
        var existing = _userRepository.FindByEmail(email);
        if (existing != null)
        {
            throw new ValidationError("email already registered");
        }
        var user = new User(id, email, name);
        _userRepository.Save(user);
        return user;
    }

    public void PromoteToAdmin(int userId)
    {
        var user = _userRepository.FindById(userId);
        Validator.RequirePositive(userId, "userId");
        user.PromoteToAdmin();
        _userRepository.Save(user);
    }
}
